### ISSUE TITLE
[TypeScript] phaser.d.ts incompatible with new flag --strictNullChecks (TS 2.0.0) 

### ISSUE THREAD

**Issue by Majirefy** [photonstorm/phaser-ce#174]
This Issue is about:
An error in the TypeScript defs
This issue can be referred to the one in original v2 phaser repo:
https://github.com/photonstorm/phaser/issues/2673

When adding strictNullChecks flag in `tsconfig.js`, it comes errors:

```
../Phaser/typescript/phaser.d.ts(2529,11): error TS2415: Class 'Phaser.Matrix' incorrectly extends base class 'PIXI.Matrix'.
  Types of property 'append' are incompatible.
    Type '(matrix: Matrix) => Matrix' is not assignable to type '(matrix: Matrix) => Matrix'.
      Types of parameters 'matrix' and 'matrix' are incompatible.
        Type 'PIXI.Matrix' is not assignable to type 'Phaser.Matrix'.
          Property 'type' is missing in type 'Matrix'.
../Phaser/typescript/phaser.d.ts(4167,11): error TS2415: Class 'Phaser.RenderTexture' incorrectly extends base class 'PIXI.RenderTexture'.
  Types of property 'render' are incompatible.
    Type '(displayObject: DisplayObject, matrix?: Matrix | undefined, clear?: boolean | undefined) => void' is not assignable to type '(displayObject: DisplayObject, matrix?: Matrix | undefined, clear?: boolean | undefined) => void'.

      Types of parameters 'matrix' and 'matrix' are incompatible.
        Type 'Matrix | undefined' is not assignable to type 'Matrix | undefined'.
          Type 'Matrix' is not assignable to type 'Matrix | undefined'.
            Type 'PIXI.Matrix' is not assignable to type 'Phaser.Matrix'.
```

**Comment by samme** [comment]
Awaiting resolution of <https://github.com/Microsoft/TypeScript/issues/10150>

**Comment by Majirefy** [comment]
@samme It seems there is no solution?
Details:https://github.com/Microsoft/TypeScript/issues/10150

Any ideas on it?

**Comment by samme** [comment]
I don't know TypeScript but it's possible that since 2.7.0 Phaser.Matrix and Phaser.RenderTexture no longer extend their PIXI counterparts.

**Comment by photonstorm** [comment]
Even if they still did it would be a quite easy change to remove the Pixi namespace. May make things like webpack bundling easier too.

**Comment by Dreaded-Gnu** [comment]
`PIXI.Matrix` will be overwritten with `Phaser.Matrix`, so the extend within typescript there can be removed completely:
https://github.com/photonstorm/phaser-ce/blob/29bc2b2d76f432a34fdb36ecc8ee634e0d8fdaf0/src/geom/Matrix.js#L388

Also having a look at `Phaser.RenderTexture` the extends within the typings seems to be completely wrong. `Phaser.RenderTexture` extends `PIXI.Texture`:
https://github.com/photonstorm/phaser-ce/blob/29bc2b2d76f432a34fdb36ecc8ee634e0d8fdaf0/src/gameobjects/RenderTexture.js#L115

**Comment by Majirefy** [comment]
@photonstorm So removal of PIXI namespace will work?
How to do that?

**Comment by photonstorm** [comment]
If you search for `PIXI.Matrix` and `PIXI.identityMatrix` in the source you'll find only a handful of instances where it is used. So I would say it's a very easy task to replace those with the Phaser version, then you can remove it from the Pixi namespace entirely.

My comment above however was implying that it would be a worthwhile endeavour to remove the PIXI namespace _entirely_, and not use it anywhere in the source. It would create a much more level playing field.

**Comment by Majirefy** [comment]
@photonstorm Thanks!
Maybe it's beyond my ability.

**Comment by samme** [comment]
Is there any way I can reproduce this without making a new TS project? I don't get errors when running `grunt tsdocs` on the source.

**Comment by rblopes** [comment]
Sorry, doesn't Phaser still depends on TypeScript 1.4.1 in its Grunt tasks? Anyway, I tried for curiosity upgrading (with `yarn upgrade-interactive`) to latest TypeScript (2.4.1), but the task stops with the following message:

```
Running "buildtsdoc:pixi" (buildtsdoc) task
Warning: Cannot read property 'text' of undefined Use --force to continue.
```

**Comment by CorayThan** [comment]
@samme I don't get it. You reference the microsoft issue:

https://github.com/Microsoft/TypeScript/issues/10150

As how this is going to be solved, but they closed that issue as a design limitation and something that won't be fixed. It sounds like photonstorm was suggesting removing the usages of the offending PIXI class. Can't that be done and this issue resolved? Why is this closed?

**Comment by samme** [comment]
It was resolved in 9e29186 and will be in the next release.

**Comment by CorayThan** [comment]
Ah, cool, I misunderstood.
